File: python/gamex/families/Gamebryo/formats/binary.py
import os
from io import BytesIO
from enum import Enum
from openstk.core import IWriteToStream
from gamex import FileSource, ArcBinaryT, MetaManager, MetaInfo, MetaContent, IHaveMetaInfo, DesSer
from gamex.families.Gamebryo.formats.nif import NiReader, NiSkinInstance, NiSourceTexture

# typedefs
class BinaryReader: pass
class Archive: pass
class BinaryArchive: pass

#region Binary_Nif

# Binary_Nif
class Binary_Nif(NiReader, IHaveMetaInfo, IWriteToStream):
    @staticmethod
    async def factory(r: BinaryReader, f: FileSource, s: Archive): return Binary_Nif(r, f)

    def __init__(self, r: BinaryReader, f: FileSource):
        super().__init__(r)
        self.name = os.path.splitext(os.path.basename(f.path))[0]

    def writeToStream(self, stream: object): return DesSer.serialize(self, stream)

    def __repr__(self): return DesSer.serialize(self)

    #region IModel

    def create(platform: str, func: callable):
        raise NotImplementedError('create')

    #endregion

    def isSkinnedMesh(self) -> bool: return any((s for s in self.blocks if isinstance(s, NiSkinInstance)))

    def getTexturePaths(self) -> list[str]: return (s.fileName for s in self.blocks if isinstance(s, NiSourceTexture) and s.fileName) 

    def getInfoNodes(self, resource: MetaManager = None, file: FileSource = None, tag: object = None) -> list[MetaInfo]: return [
        MetaInfo(None, MetaContent(type = 'Text', name = os.path.basename(file.path), value = self)),
        MetaInfo('NIF', items = [
            MetaInfo(f'NumBlocks: {self.numBlocks}')
            ])
        ]

#endregion

File: dotnet/Families/GameX.Gamebryo/Formats/Binary.cs
using GameX.Gamebryo.Formats.Nif;
using OpenStack.Gfx;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace GameX.Gamebryo.Formats;

#region Binary_Nif

public class Binary_Nif(BinaryReader r, FileSource f) : NiReader(r), IHaveMetaInfo, IModel, IWriteToStream {
    public static Task<object> Factory(BinaryReader r, FileSource f, Archive s) => Task.FromResult((object)new Binary_Nif(r, f));

    public string Name = Path.GetFileNameWithoutExtension(f.Path);

    #region IModel

    public T Create<T>(string platform, Func<object, T> func) {
        //Activator.CreateInstance("");
        func(null);
        return default;
    }

    #endregion

    public bool IsSkinnedMesh() => Blocks.Any(s => s is NiSkinInstance);

    public IEnumerable<string> GetTexturePaths() => Blocks.Select(s => s is NiSourceTexture z && !string.IsNullOrEmpty(z.FileName) ? z.FileName : null).Where(s => s != null);

    List<MetaInfo> IHaveMetaInfo.GetInfoNodes(MetaManager resource, FileSource file, object tag) => [
        new(null, new MetaContent { Type = "Object", Name = Name, Value = this }),
        new("NIF", items: [
            new($"NumBlocks: {NumBlocks}"),
        ]),
    ];

    public void WriteToStream(Stream stream) => this.Serialize(stream);

    public override string ToString() => this.Serialize();
}

#endregion